lib/mongoid/fields.rb

```diff
@@ -241,6 +241,28 @@ def using_object_ids?
       self.class.using_object_ids?
     end
 
+    # Does this field start with a dollar sign ($) or contain a dot/period (.)?
+    #
+    # @param [ String ] name The field name.
+    #
+    # @return [ true, false ] If this field is dotted or dollared.
+    def dot_dollar_field?(name)
+      n = aliased_fields[name] || name
+      fields.key?(n) && (n.include?('.') || n.start_with?('$'))
```

Comment:
Can you explain why the `key?` condition is here please? Wouldn't it allow writes to dotted fields that aren't already persisted?